datasync/chaindatafetcher/chaindata_fetcher.go

```diff
@@ -0,0 +1,135 @@
+package chaindatafetcher
+
+import (
+	"github.com/klaytn/klaytn/blockchain"
+	"github.com/klaytn/klaytn/event"
+	"github.com/klaytn/klaytn/log"
+	"github.com/klaytn/klaytn/networks/p2p"
+	"github.com/klaytn/klaytn/networks/rpc"
+	"github.com/klaytn/klaytn/node"
+	"sync"
+)
+
+var logger = log.NewModuleLogger(log.ChainDataFetcher)
+
+type ChainDataFetcher struct {
+	config *ChainDataFetcherConfig
+
+	blockchain *blockchain.BlockChain
+
+	chainCh  chan blockchain.ChainEvent
+	chainSub event.Subscription
+
+	reqCh  chan uint64 // TODO-ChainDataFetcher add logic to insert new requests from APIs to this channel
+	resCh  chan uint64
+	stopCh chan struct{}
+
+	numHandlers int
+
+	wg sync.WaitGroup
+}
+
+func NewChainDataFetcher(ctx *node.ServiceContext, cfg *ChainDataFetcherConfig) (*ChainDataFetcher, error) {
+	return &ChainDataFetcher{
+		config:      cfg,
+		chainCh:     make(chan blockchain.ChainEvent, cfg.BlockChannelSize),
+		reqCh:       make(chan uint64, cfg.JobChannelSize),
+		resCh:       make(chan uint64, cfg.JobChannelSize),
+		stopCh:      make(chan struct{}),
+		numHandlers: cfg.NumHandlers,
+	}, nil
+}
+
+func (f *ChainDataFetcher) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{}
+}
+
+func (f *ChainDataFetcher) APIs() []rpc.API {
+	// TODO-ChainDataFetcher add APIs to start or stop chaindata fetcher
+	return []rpc.API{}
+}
+
+func (f *ChainDataFetcher) Start(server p2p.Server) error {
+	// launch multiple goroutines to handle new blocks
+	for i := 0; i < f.numHandlers; i++ {
+		go f.handler()
+	}
+
+	// subscribe chain head event
+	f.chainSub = f.blockchain.SubscribeChainEvent(f.chainCh)
+	go f.reqLoop()
+	go f.resLoop()
+
+	return nil
+}
+
+func (f *ChainDataFetcher) Stop() error {
+	f.chainSub.Unsubscribe()
+	close(f.stopCh)
+
+	logger.Info("wait for all goroutines to be terminated...")
+	f.wg.Wait()
+	logger.Info("terminated all goroutines for chaindatafetcher")
+	return nil
+}
+
+func (f *ChainDataFetcher) Components() []interface{} {
+	return nil
+}
+
+func (f *ChainDataFetcher) SetComponents(components []interface{}) {
+	for _, component := range components {
+		switch v := component.(type) {
+		case *blockchain.BlockChain:
+			f.blockchain = v
+		}
+	}
+}
+
+func (f *ChainDataFetcher) handler() {
```

Comment:
Please, consider renaming this method. `handler` sounds like returning handlers, but actually generates/adds a handler.

Reply:
Applied. c7bb50563296063b05fd4c487934232bb06aa0ec